src/hooks/mongodb4x.ts

```diff
@@ -32,26 +48,51 @@ const injectMonitoringCommand = (args: any[]) => {
           );
       }
   }
+  return args;
 };
 
-export const beforeConstructorHook = (args: any[], extenderContext: any) => {
-  injectMonitoringCommand(args);
-};
-
-const isMonitoringEnabled = (args: any[]) => {
-  return args.length > 1 && typeof args[1] === 'object' && args[1].monitorCommands === true;
-};
-
-export const afterConstructorHook = (args: any[], clientInstance: any, extenderContext: any) => {
-  if (isMonitoringEnabled(args)) {
-    try {
-      clientInstance.on('commandStarted', safeExecute(onStartedHook));
-      clientInstance.on('commandSucceeded', safeExecute(onSucceededHook));
-      clientInstance.on('commandFailed', safeExecute(onFailedHook));
-    } catch (err) {
-      logger.warn(`MongoDB 4.x 'on' hooks cannot be applied to ${clientInstance}`, err);
-    }
-  } else {
-    logger.debug("MongoDB 4.x 'on' hooks skipped: monitorCommands was not set in the options");
+export const wrapMongoClient4xClass = (mongoClientLibrary: any) => {
+  try {
+    const originalStaticConnect = mongoClientLibrary.MongoClient.connect;
+    const staticProperties = getStaticProperties(mongoClientLibrary.MongoClient);
+    const wrapper = (originalFn: any) => {
+      const wrappedClass = function (...args: any[]) {
+        const client = new originalFn(...injectMonitoringCommand(args));
+        attachEventHooks(client);
+        return client;
+      };
+      // ensure that we don't lose any unexpected static properties
+      for (const key of staticProperties) {
+        wrappedClass[key] = originalFn[key];
+      }
+      // wrap the original connect method. this references the original constructor
+      // and not our overridden one, so we must attach the hooks to the resulting
+      // client instance
+      wrappedClass.connect = (url: string, options: any, callback: Function) => {
+        return new Promise((resolve, reject) => {
+          try {
+            callback =
+              typeof callback === 'function'
+                ? callback
+                : typeof options === 'function'
+                ? options
+                : undefined;
```

Comment:
Please don't chain ternary operators. This operator should be used only for simple cases.

Reply:
Lets leave it for now